walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -21,51 +21,112 @@
+      }
     }
 
     /*
-      If there are no times that the meeting cannot happen, available all day.
-      this will only happen if the person invited to the meeting is not invited
-      to the event(s).
+      If timeRanges and optionalTimeRanges are empty, that means neither
+      attendees or optional attendees have any other events scheduled for the day,
+      meaning they are free all day for the meeting
     */
-    if (timeRanges.isEmpty()) {
+    if (timeRanges.isEmpty() && optionalTimeRanges.isEmpty()) {
       meetingTimes.add(TimeRange.WHOLE_DAY);
       return meetingTimes;
     }
 
+    /*
+      if timeranges is empty but optionalTimeRanges isnt, this means that 
+      the only restrictions to time will come from optional attendees, which
+      means they are the only ones that need to be considered when building
+      an array of available times
+    */
+    if (timeRanges.isEmpty() && !optionalTimeRanges.isEmpty()) {
+      Collections.sort(optionalTimeRanges, TimeRange.ORDER_BY_START);
+      optionalTimeRanges = findOverlap(optionalTimeRanges);
+      optionalTimeRanges = findMeetingTime(optionalTimeRanges, duration);
+      return optionalTimeRanges;
+    }
+
+    /*
+      if timeRanges isn't empty but optionalTimeRanges is, that means that 
+      the only restrictions to time will come from mandatory attendees, which
+      means they are the only ones that ened to be considered when building
+      an array of available times
+    */
+    if (!timeRanges.isEmpty() && optionalTimeRanges.isEmpty()) {
+      Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+      timeRanges = findOverlap(timeRanges);
+      timeRanges = findMeetingTime(timeRanges, duration);
+      return timeRanges;
+    }
+
     //sort the time ranges by start time in ascending order 
     Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    Collections.sort(optionalTimeRanges, TimeRange.ORDER_BY_START);
 
     //check if time range containes the time range after it, if so
     //combine them, if not continue on
-    for (int i = 0; i < timeRanges.size() - 1; ++i) {
+    timeRanges = findOverlap(timeRanges);
+    optionalTimeRanges = findOverlap(optionalTimeRanges);
+
+    /*
+      create array of time ranges that are available by filling in the
+      gaps in timeRanges (which holds all the times a meeting can't be).
+      Checks to make sure each time range fits the specified duration
+    */
+    timeRanges = findMeetingTime(timeRanges, duration);
+    optionalTimeRanges = findMeetingTime(optionalTimeRanges, duration);
+
+    return filterTimeRanges(timeRanges, optionalTimeRanges, duration);
+  }
+
+  //check if time range containes the time range after it, if so
+  //combine them, if not continue on
+  private List<TimeRange> findOverlap(List<TimeRange> timeRanges) {
+    for (int i = 0; i < timeRanges.size() - 1;) {
+      if(timeRanges.get(i).end() == timeRanges.get(i+1).start()) {
```

Comment:
NIT: spaces between operators, same for following lines
https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace